### PR TITLE
SALTO-2124: add restriction filter for ticket_field__custom_field_options.value

### PR DESCRIPTION
**Pull request by edenhassid**
_Add restriction filter for ticket_field__custom_field_options.value_

---
_Additional context for reviewer_:
None

---
_Release Notes_: 
_Zendesk adapter_
* Add restriction filter of regex for the type ticket_field__custom_field_options.value

---
_User Notifications_: 
_Zendesk adapter_
* Add restriction filter of regex for the type ticket_field__custom_field_options.value

### CODE REVIEW

File: packages/zendesk-adapter/src/filters/add_restriction.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+} from '@salto-io/adapter-api'
+import { FilterCreator } from '../filter'
+
+type AllRestrictionsToMake = Record<string, RestrictionAnnotationType>
+type RestrictionByType = Record<string, AllRestrictionsToMake>
+
+const TO_ADD_RESTRICTION: RestrictionByType = {
+  ticket_field__custom_field_options: {
+    value: {
+      regex: '^[0-9A-Za-z-_.\\/~:^]+$',
```
Comment (noamhammer) [inline packages/zendesk-adapter/src/filters/add_restriction.ts:32]: I think that you should also add `enforce_value: true` here, no?
read about all the restrictions options [here](https://github.com/salto-io/salto/blob/main/docs/syntax.md#_restriction)

File: packages/zendesk-adapter/src/filters/add_restriction.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  isObjectType,
+  ObjectType,
+  RestrictionAnnotationType,
+} from '@salto-io/adapter-api'
+import { FilterCreator } from '../filter'
+
+type AllRestrictionsToMake = Record<string, RestrictionAnnotationType>
+type RestrictionByType = Record<string, AllRestrictionsToMake>
+
+const TO_ADD_RESTRICTION: RestrictionByType = {
```
Comment (noamhammer) [inline packages/zendesk-adapter/src/filters/add_restriction.ts:29]: nit: maybe rename to TYPE_NAME_TO_FIELD_RESTRICTIONS or something like that?

File: packages/zendesk-adapter/src/filters/add_restriction.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      regex: '^[0-9A-Za-z-_.\\/~:^]+$',
+    },
+  },
+}
+
+
+const addRestriction = (obj:ObjectType) : void => {
+  const typeToChange = TO_ADD_RESTRICTION[obj.elemID.typeName]
+  Object.keys(typeToChange).forEach(field => {
+    if (obj.fields[field]?.annotations) {
```
Comment (noamhammer) [inline packages/zendesk-adapter/src/filters/add_restriction.ts:41]: I think that you should change this if to be `if (Object.prototype.hasOwnProperty.call(typeToChange, field)) {`
For more info about why we do hasOwnPropery, please read [this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty#direct_vs._inherited_properties)

File: packages/zendesk-adapter/src/filters/add_restriction.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      )
+    }
+  })
+}
+
+const filterCreator: FilterCreator = () => ({
+  onFetch: async (elements: Element[]): Promise<void> => {
+    elements
+      .filter(isObjectType)
+      .filter(obj => TO_ADD_RESTRICTION[obj.elemID.typeName] ?? false)
```
Comment (noamhammer) [inline packages/zendesk-adapter/src/filters/add_restriction.ts:53]: maybe `.filter(obj => Object.keys(TO_ADD_RESTRICTION).includes(obj.elemID.typeName))`?

File: packages/zendesk-adapter/test/filters/add_restriction.test.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        expect(elements[0].fields.value.annotations[CORE_ANNOTATIONS.RESTRICTION].regex)
+          .toBeDefined()
+        expect(elements[0].fields.value.annotations[CORE_ANNOTATIONS.RESTRICTION].regex)
+          .toEqual('^[0-9A-Za-z-_.\\/~:^]+$')
+      })
+      it('should not add a regex restriction to the value field when the field does not exists', async () => {
+        const elements = [objTypeNoValue].map(e => e.clone())
+        await filter.onFetch(elements)
+        expect(elements).toHaveLength(1)
+        // eslint-disable-next-line max-len
```
Comment (noamhammer) [inline packages/zendesk-adapter/test/filters/add_restriction.test.ts:65]: please don't use this. Just split the long line

File: packages/zendesk-adapter/test/filters/add_restriction.test.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+          client,
+          paginationFuncCreator: paginate,
+        }),
+        config: DEFAULT_CONFIG,
+        fetchQuery: elementUtils.query.createMockQuery(),
+      }) as FilterType
+    })
+
+    describe('onFetch', () => {
+      it('should add a regex restriction to the value field when exists', async () => {
```
Comment (noamhammer) [inline packages/zendesk-adapter/test/filters/add_restriction.test.ts:52]: maybe add a test for a type with two fields - one has a restiction and the other doesn't and check that you get what you expect?

**Review by noamhammer — APPROVED** [review]

File: packages/zendesk-adapter/src/filters/add_restriction.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  ticket_field__custom_field_options: {
+    value: {
+      regex: '^[0-9A-Za-z-_.\\/~:^]+$',
+      enforce_value: true,
+    },
+  },
+}
+
+
+const addRestriction = (obj: ObjectType) : void => {
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/add_restriction.ts:39]: @noamhammer @edenhassid I might be missing something but couldn't we add this with the infra (like [here](https://github.com/salto-io/salto/blob/main/packages/zendesk-adapter/src/config.ts#L492))? why did we need a filter?

**Pull request merged**